[app] Add Header Component

zerefwayne opened the pull request:
**Description**

- This PR adds a header component to the app.

- A core module is added to the project which will contain all the business logic which is not related to displaying content. This will contain all the services, pipes and directives that will be used globally. This module is imported in `app.module`.

- To manage the theme throughout the app, a new `theme.service` has been created inside the core module. The header component invokes the `toggleTheme` function in the service which toggles the theme and updates all the components (`app.component` for now) of the theme change. `app.component` has subscribed to the Event Emitter of the theme service and listens for the updates.

**Screenshot**


https://user-images.githubusercontent.com/28949397/123653903-2e8be900-d84b-11eb-9ce1-9d9fb9e69f2a.mp4


**To be merged after #4**

zerefwayne commented on diagnostics-app/src/app/app.component.unit.spec.ts:
```diff
@@ -30,28 +32,7 @@ describe('component: app', () => {
-    fixture.detectChanges();
-    expect(component.toggleTheme).toHaveBeenCalled();
+    // default cssClass is dark
     expect(component.cssClass).toBe('dark');
```
Test is no longer required here.

zerefwayne commented on diagnostics-app/src/app/layout/header/header.component.ts:
```diff
@@ -0,0 +1,40 @@
+export class HeaderComponent {
+  private _theme: Theme;
+
+  get theme() {
+    return this._theme;
+  }
```
We need to keep a private variable because reactivity is absent between service and component classes.

MahmoudAGawad commented on diagnostics-app/src/app/app.component.ts:
```diff
@@ -8,32 +8,29 @@
-  get theme() {
-    return this._theme;
-  }
+  @HostBinding('class') private _cssClass!: Theme;
```
What is the added value from changing string to Theme?

zerefwayne replied:
Same as below.

MahmoudAGawad commented on diagnostics-app/src/app/app.module.ts:
```diff
@@ -9,15 +9,19 @@
 import { environment } from '../environments/environment';
 import { BrowserAnimationsModule } from '@angular/platform-browser/animations';
 
+import { SharedModule } from './shared/shared.module';
+
+import { AppComponent } from './app.component';
+import { HeaderComponent } from './layout/header/header.component';
+import { CoreModule } from './core/core.module';
```
Curious why you moved import around and if there is a best practice for import ordering?

zerefwayne replied:
Re-arranged the imports. The best practice is to have all @angular imports on top, then the modules created inside the project and then the components.

zerefwayne replied:
https://angular.io/guide/component-interaction#parent-and-children-communicate-using-a-service

According to this, it is much better to use `rxjs/Subject` rather than `EventEmitter`. I think this solves the purpose of abstracting the `emit` method to a function and provide only the `subscribe` object as public.

MahmoudAGawad replied:
Great. Let's do this in the other files as well.

MahmoudAGawad commented on diagnostics-app/src/app/core/services/theme.service.ts:
```diff
@@ -0,0 +1,36 @@
+export class ThemeService {
+  private _theme: Theme = 'dark';
+
+  // subscribed by app and header
+  // to listen for changes
```
should be in one line.

MahmoudAGawad commented on diagnostics-app/src/app/core/services/theme.service.ts:
```diff
@@ -0,0 +1,36 @@
+
+  // subscribed by app and header
+  // to listen for changes
+  themeUpdated: EventEmitter<Theme> = new EventEmitter();
```
I think ThemeService should take ownership of (un)subscribing events.

EventEmitter should be an implementation detail (that we might change it in the future). So in ThemeService, you would expose (un)subscribe() methods to clients and use an instance of EventEmitter to do the actual work (subscribing).

MahmoudAGawad replied:
nit: "themeEventHandler" may be a better name than "themeUpdated".

zerefwayne replied:
https://angular.io/guide/component-interaction#parent-and-children-communicate-using-a-service

According to this, it is much better to use `rxjs/Subject` rather than `EventEmitter`. I think this solves the purpose of abstracting the `emit` method to a function and provide only the observable stream object as public.

zerefwayne replied:
I've replaced EventEmitter with the Subject like Angular. Now components are not allowed to emit anything, they can only subscribe to the observable exposed by the service.

MahmoudAGawad commented on diagnostics-app/src/app/core/services/theme.service.ts:
```diff
@@ -0,0 +1,36 @@
+  providedIn: 'root',
+})
+export class ThemeService {
+  private _theme: Theme = 'dark';
```
Again: I think It doesn't add anything for the type to be Theme rather than string. In fact, you added a new enum file :)

zerefwayne replied:
I used a type because I was using `'dark' | 'light'` at a lot of places. It checks the value at every place where we use theme and provides a nice auto-complete as well. Thus, using the benefits TypeScript provides.

zerefwayne replied:
I had a look at TypeScript best practices and updated the project structure accordingly. There's a `types` directory inside `core` and all the types that will be used by various modules/components are added in `global.types.ts`. All the types we create later will reside in this directory.

Similarly when we need `interfaces`, they'll be added in `core/interfaces`.

MahmoudAGawad replied:
> I used a type because I was using `'dark' | 'light'` at a lot of places.

Can you mention where those places are?

MahmoudAGawad replied:
Please refer to my comment in global.types.ts.

I think what you are looking for here is Enum and not typealias. By using enums, you can refer to its properties as Theme.DARK or Theme.LIGHT, and forget about using 'dark' and 'light' strings.

zerefwayne replied:
Yes, I was trying to limit the values. According to multiple typescript sources, [best one](https://blog.bam.tech/developer-news/should-you-use-enums-or-union-types-in-typescript): 

> This means using union types by default, and maybe making an exception if you like to rename things all the time or your tooling doesn’t provide live typecheking and code completion.

A union type seems to be a better fit here since it has to be a string type with limited values that are not going to change anytime soon.

[union type vs enum](https://thisthat.dev/literal-union-type-vs-string-enums/)

MahmoudAGawad commented on diagnostics-app/src/app/app.component.it.spec.ts:
```diff
@@ -0,0 +1,68 @@
+// found in the LICENSE file.
+
+/**
+ * @fileoverview Test file for the app.component.ts
```
nit: Integration Test*

MahmoudAGawad commented on diagnostics-app/src/app/app.component.it.spec.ts:
```diff
@@ -0,0 +1,68 @@
+    // subscribe to changes from themeService
+    // expect statements will run on receiving toggleUpdates
+    themeService.themeEventHandler.subscribe((next: Theme) => {
+      if (currentClass === 'dark') {
+        // gets executed when currentClass is dark
+        expect(next).toBe('light');
+      } else {
+        // gets executed when currentClass is light
+        expect(next).toBe('dark');
+      }
```
It seems off to include logic in a test (even when it is simple). I think it is better to split the test into two (change dark mode to light mode, and vice versa). This will require setting the theme as a pre-test-work.

Or even better, to use parameterized test (if Angular supports it): https://rjzaworski.com/2013/01/parameterized-testing-in-javascript?

WDYT?

zerefwayne replied:
Yeah it seemed off to me as well. This is the only way to test both cases inside a single function call. I went through the blog and yes, this is supported by Jasmine + Karma. I think when we have complicated logic later in the project, we can use parametrized testing.

For now, I think splitting this one into two tests makes sense because this is a simple one.

MahmoudAGawad commented on diagnostics-app/src/app/app.component.it.spec.ts:
```diff
@@ -0,0 +1,68 @@
+
+    // sets current class 'light'
+    currentClass = component.cssClass;
+    themeService.toggleTheme();
+    fixture.detectChanges();
+    // class should change to dark
+    expect(component.cssClass).toBe('dark');
```
we don't need to toggleTheme() twice in the same test. It is enough to test that toggleTheme() "works"

zerefwayne replied:
Okay, I have done this in multiple places. So I'll remove the redundant `expect`s from the tests. And only test `toggleTheme()` inside the service properly.

MahmoudAGawad commented on diagnostics-app/src/app/core/types/global.types.ts:
```diff
@@ -0,0 +1,9 @@
+ * @fileoverview Defines types used in global scope by multiple modules.
+ */
+
+export type Theme = 'dark' | 'light';
```
IIUC, typeof Theme is still a string. It seems like you are trying to limit the assignable values by doing this which you can achieve better using enums.

MahmoudAGawad commented on diagnostics-app/src/app/core/services/theme.service.ts:
```diff
@@ -0,0 +1,37 @@
+  private themeChangedSource = new Subject<Theme>();
+
+  // subscribed by app and header to listen for changes
+  public themeChanged$ = this.themeChangedSource.asObservable();
```
In a previous comment, you used EventEmitter, which along with Subject are implementation details.

It is very probable to change this implementation afterwards anytime (you just changed this just now from EventEmitter to Subject), and this change shouldn't be visible to the user of this service. That's why the users/clients should not have access to this instance, e.g. themeService.themeChanged$.subscribe(...), instead they depend on the ThemeService stable interface, themeService.subscribe(.....).

MahmoudAGawad commented on diagnostics-app/src/app/layout/header/header.component.it.spec.ts:
```diff
@@ -0,0 +1,68 @@
+// found in the LICENSE file.
+
+/**
+ * @fileoverview Integration Tests for header component.
```
nit: "Integration Tests" -> "Integration tests"
you used small case in the unit test file. Just want to be consistent.

zerefwayne commented:
@MahmoudAGawad I tried experimenting with creating a `themeChanged(cb)` method in `theme.service`. The subscribe part works well. However, there is a major issue with unsubscribing this callback when component is destroyed. If we do not call an unsubscribe method, the callback function lives in memory and keeps receiving the updates.

To resolve this issue, the `themeChanged(cb)` must return an `rxjs/Subscription` object to the component. Using this we can unsubscribe the subscription in `ngOnDestroy` lifecycle hook.

**theme.service.ts**

```ts
import { Subject, Subscription } from 'rxjs';
...
private _theme: Theme = 'dark';
private themeChangedSource = new Subject<Theme>();
...
public themeChanged(cb: (next: Theme) => void): Subscription {
    return this.themeChangedSource.subscribe(cb);
}
...
```

**app.component.ts**

```ts
import { Subscription } from 'rxjs';
...
private csSubscription!: Subscription;

ngOnInit() {
    this._cssClass = this.themeService.theme;
    // listen to theme updates from themeService
    this.csSubscription = this.themeService.themeChanged((nextTheme: Theme) => {
      this._cssClass = nextTheme;
    });
}

ngOnDestroy() {
    this.csSubscription.unsubscribe();
}

```
But this defeats our purpose of abstracting away the implementation details because we have to call the unsubscribe method anyways.

Apart from this, when we expose the Observable `themeChanged$` directly to the components, apart from the`subscribe` method, they can access a lot of useful methods that come along with `rxjs` such as pipe(), map(), [etc.](https://www.learnrxjs.io/learn-rxjs/operators)

____

If we compare this to the approach where we expose the observable, the pros are:

1) Angular handles unsubscribe automatically for us; thus we avoid the hassle of storing the subscription and manually unsubscribing it.
2) We get access to these operators that we might need later on when we use complex observables that receive live data from the extension.

Handling the subscription inside theme service will give us the advantage of hiding the detail only in the subscription part, we'll have to anyways manually expose `rxjs/Subscription` for components to unsubscribe from. Since rxjs is a popular library and is provided out of the box in Angular, I think exposing a safe readonly object is more preferrable as compared to hiding it away.

MahmoudAGawad commented:
> @MahmoudAGawad I tried experimenting with creating a `themeChanged(cb)` method in `theme.service`. The subscribe part works well. However, there is a major issue with unsubscribing this callback when component is destroyed. If we do not call an unsubscribe method, the callback function lives in memory and keeps receiving the updates.

Makes sense.

> 
> To resolve this issue, the `themeChanged(cb)` must return an `rxjs/Subscription` object to the component. Using this we can unsubscribe the subscription in `ngOnDestroy` lifecycle hook.
> 
> **theme.service.ts**
> 
> ```ts
> import { Subject, Subscription } from 'rxjs';
> ...
> private _theme: Theme = 'dark';
> private themeChangedSource = new Subject<Theme>();
> ...
> public themeChanged(cb: (next: Theme) => void): Subscription {
>     return this.themeChangedSource.subscribe(cb);
> }
> ...
> ```
> 
> **app.component.ts**
> 
> ```ts
> import { Subscription } from 'rxjs';
> ...
> private csSubscription!: Subscription;
> 
> ngOnInit() {
>     this._cssClass = this.themeService.theme;
>     // listen to theme updates from themeService
>     this.csSubscription = this.themeService.themeChanged((nextTheme: Theme) => {
>       this._cssClass = nextTheme;
>     });
> }
> 
> ngOnDestroy() {
>     this.csSubscription.unsubscribe();
> }
> ```

This approach is good. Couple of comments:
- nit: themeChanged() --> subscribeOnThemeChange()
- In theme.service.ts we can augment the implementation to use refCount (https://rxjs.dev/guide/subject#reference-counting) to make the Observable stop executing when no observers are there.

> 
> But this defeats our purpose of abstracting away the implementation details because we have to call the unsubscribe method anyways.

I think this is Ok. As you mentioned, Subject/Subscription is a good angular abstraction.

> 
> Apart from this, when we expose the Observable `themeChanged$` directly to the components, apart from the`subscribe` method, they can access a lot of useful methods that come along with `rxjs` such as pipe(), map(), [etc.](https://www.learnrxjs.io/learn-rxjs/operators)
> 
> If we compare this to the approach where we expose the observable, the pros are:
> 
> 1. Angular handles unsubscribe automatically for us; thus we avoid the hassle of storing the subscription and manually unsubscribing it.

Are you sure that Angular handles unsubscription automatically? If so, when? does it call unsubscribe automatically? what is the difference between this and maintaining Subscription object?

> 2. We get access to these operators that we might need later on when we use complex observables that receive live data from the extension.

1- What does "receiving live data from the extension" have to do with the Theme state change?
2- Subject is a special type of Observable, it inhereits all peroperties.

> 
> Handling the subscription inside theme service will give us the advantage of hiding the detail only in the subscription part, we'll have to anyways manually expose `rxjs/Subscription` for components to unsubscribe from. Since rxjs is a popular library and is provided out of the box in Angular, I think exposing a safe readonly object is more preferrable as compared to hiding it away.

I wonder why you use Observable (Subject.asObservable()) in the first place? The difference between Subject and Observable is mentioned in the [Documentation](https://rxjs.dev/api/index/class/Subject), that's Subject multicasts values to many observers, while Observables are unicast.
Looking at Theme state change use-case, when the theme is changed, we want to inform components about the change, this seems multicast to me.
That said, It seems off that you change a Subject to Observable (Subject.asObservable). In fact, the opposite is fine, Observable -> Subject -> (many observers), see this: https://rxjs.dev/guide/subject#multicasted-observables.

Observable vs Subject: In theory, it shouldn't matter whether you use Subject or Observable, we have to put unsubscription logic (i think) anyway. However, Subject is better suited here to multicast theme changes to multiple components (observers). Each component need to a way to subscribe and unsubscribe on the values. Ideally, unsubscribe should be via a Subscription object.

--

Taking a step back, I think the notion of observerable/observer doesn't fit the theme change use case. We use the observer pattern to keep track of values over time (the Subject interface has next() method which indicates just that), take the CPU utilization chart example, where the drawer needs to observe CPU utilization values over time to draw them on chart. Components like main app are "NOT" measuring/observing theme values, but rather listens on state change. I think the notion here should be, a theme change triggers an event to all interested components, so that can they take action accordingly (for the app main component, it changes the background of the main layout).

zerefwayne committed:
[app] Add header component

[app] Refactor updates

[app] Inject defaultTheme parameter to constructor

[app] Update header tests

[app] Remove redundant expects

[app] Remove 'private'

[app] Add comments

[app] Refactor nits

MahmoudAGawad commented on diagnostics-app/src/app/app.component.it.spec.ts:
```diff
@@ -0,0 +1,50 @@
+    themeService = TestBed.inject(ThemeService);
+  });
+
+  it('set cssClass on receiving update from themeService', () => {
```
at the beginning, we should expect the theme to be DARK so that in line 47 we make sure it was changed.

zerefwayne replied:
Yes it is by default Theme.DARK

MahmoudAGawad replied:
We both know this, but the test does not :))

zerefwayne replied:
Added an expect statement.

MahmoudAGawad commented on diagnostics-app/src/app/app.component.it.spec.ts:
```diff
@@ -0,0 +1,50 @@
+  });
+
+  it('set cssClass on receiving update from themeService', () => {
+    // subscribe to changes from themeService
+    // expect statements will run on receiving toggleUpdates
+    let subscription = themeService.subscribeOnThemeChange((next: Theme) => {
+      expect(next).toBe(Theme.LIGHT);
+    });
```
why do you want to add another subscriber? IIUC, this test tests that app.component.ts is able to subscribe on themeChange and we verify that in line 47.

Also there is no guarantee that this callback is actually called, i.e. if this callback is not called the test can be green, regardless.

zerefwayne replied:
Yes, just verifying correct cssClass can do the job. 👍🏻

MahmoudAGawad commented on diagnostics-app/src/app/app.component.ts:
```diff
@@ -8,32 +8,36 @@
+    this._cssClass = this._themeService.theme;
+    // listen to theme updates from themeService
+    this._themeSubscription = this._themeService.subscribeOnThemeChange(
+      (nextTheme: Theme) => {
+        this._cssClass = nextTheme;
+      }
```
nit: WDYT about renaming nextTheme -> newTheme? "next" indicates interest in past data points.

MahmoudAGawad commented on diagnostics-app/src/app/core/enums/global.enums.ts:
```diff
@@ -0,0 +1,12 @@
+// Copyright 2021 The Chromium Authors. All rights reserved.
```
regarding file name: do you plan to only define enums here? or can include general types as well? if the latter, then we can rename it global.types.ts. Up to you.

zerefwayne replied:
Yeah mostly we will have two folders `enums` and `interfaces`. I'll add the `interfaces` folder when we need them.

MahmoudAGawad commented on diagnostics-app/src/app/core/services/theme.service.ts:
```diff
@@ -0,0 +1,38 @@
+
+  constructor() {}
+
+  // getter for current theme
```
let's remove non-informative comments.

MahmoudAGawad commented on diagnostics-app/src/app/core/services/theme.service.unit.spec.ts:
```diff
@@ -0,0 +1,47 @@
+  });
+
+  it('should switch from dark to light', () => {
+    let subscription = service.subscribeOnThemeChange((next: Theme) => {
+      expect(next).toBe(Theme.LIGHT);
+    });
```
let's remove this code. The test is about toggleTheme(), before calling it we should verify the initial value (Theme,DARK), then toggleTheme(), then expect the theme to be changed.

MahmoudAGawad commented on diagnostics-app/src/app/core/services/theme.service.unit.spec.ts:
```diff
@@ -0,0 +1,47 @@
+  it('should switch from dark to light', () => {
+    // switch from dark to light before subscribing
+    service.toggleTheme();
+    let subscription = service.subscribeOnThemeChange((next: Theme) => {
+      expect(next).toBe(Theme.DARK);
+    });
```
same, let's remove this code.

MahmoudAGawad commented on diagnostics-app/src/app/core/services/theme.service.unit.spec.ts:
```diff
@@ -0,0 +1,47 @@
+    subscription.unsubscribe();
+  });
+
+  it('should switch from dark to light', () => {
```
nit: should switch from light to dark

MahmoudAGawad commented on diagnostics-app/src/app/core/services/theme.service.unit.spec.ts:
```diff
@@ -0,0 +1,47 @@
+
+  it('should switch from dark to light', () => {
+    // switch from dark to light before subscribing
+    service.toggleTheme();
```
this is the subject of this test, the test shouldn't depend that it behaves "correctly". Instead, we want to set the initial value of the Theme to be Theme.LIGHT. And this is tricky, because theme is private. One way to do this, is to call themeService constructor with the initial theme value. WDYT?

zerefwayne replied:
It was indeed tricky to pass a constructor parameter in Angular. But I found a way to do it using Angular's Dependency Injection [[example](https://www.techiediaries.com/angular-10-constructor-parameters-inject-optional/)]. 

Now in future, if needed, `defaultTheme` can be passed to the constructor using the `providers` array in `@NgModule` declaration in `app.module.ts`.

MahmoudAGawad replied:
Sounds good. Thanks for looking into this.

Using the initial value in the constructor, we can add the feature to detect the system preference and set it accordingly.

MahmoudAGawad commented on diagnostics-app/src/app/layout/header/header.component.ts:
```diff
@@ -0,0 +1,48 @@
+  }
+
+  isDarkModeActivated() {
+    return this._theme === Theme.DARK;
```
If we delegate this call to the ThemeService:
```
isDarkModeActivated() {
    return this._themeService.isDarkModeActivated();
}
```
or we directly get the them value from themeService.
```
isDarkModeActivated() {
    return this._themeService.theme === Theme.DARK;
}
```
We can git rid of maintaining _theme and _themeSubscription. We do this to leverage the observation that header component is the one initiating themeChange events (since it listens on button click).

MahmoudAGawad replied:
If you are going to choose, the delegation way is better since header component doesn't need to know about Theme.DARK.

zerefwayne replied:
Sounds great! However, the html template cannot access a method of service. I've removed everything from the component.

```
export class HeaderComponent {
  constructor(private _themeService: ThemeService) {}

  onToggleTheme() {
    this._themeService.toggleTheme();
  }

  isDarkModeActivated() {
    return this._themeService.theme === Theme.DARK;
  }
}
```

Looks much lighter and does the job. 😄 I don't think comparison is needed anywhere else except for rendering the icon. All the other styles are handled using css. So we can live the functionality here.

MahmoudAGawad replied:
Interesting. We can indeed settle with the comparison option for now.

MahmoudAGawad commented on diagnostics-app/src/app/layout/header/header.component.it.spec.ts:
```diff
@@ -0,0 +1,56 @@
+  it('should receive theme update from service on click', () => {
+    // subscribe to changes from themeService
+    // expect statements will run on receiving toggleUpdates
+    let subscription = service.subscribeOnThemeChange((next: Theme) => {
+      expect(next).toBe(Theme.LIGHT);
+    });
```
same: remove this code.

MahmoudAGawad commented on diagnostics-app/src/app/layout/header/header.component.it.spec.ts:
```diff
@@ -0,0 +1,56 @@
+  it('service should emit event on click', () => {
+    spyOn(service, 'toggleTheme');
+    button.click();
+    expect(service.toggleTheme).toHaveBeenCalled();
```
let's also verify that isDarkModeActivated() returns the correct value.

MahmoudAGawad commented on diagnostics-app/src/app/layout/header/header.component.it.spec.ts:
```diff
@@ -0,0 +1,56 @@
+    expect(service.toggleTheme).toHaveBeenCalled();
+  });
+
+  it('should receive theme update from service on click', () => {
+    // subscribe to changes from themeService
+    // expect statements will run on receiving toggleUpdates
+    let subscription = service.subscribeOnThemeChange((next: Theme) => {
+      expect(next).toBe(Theme.LIGHT);
+    });
+    // triggers the expect inside subscription
+    button.click();
+    fixture.detectChanges();
+    subscription.unsubscribe();
+  });
```
Let's remove this test entirely. The integration behavior is tested in the previous test.

MahmoudAGawad commented on diagnostics-app/src/app/core/services/theme.service.ts:
```diff
@@ -0,0 +1,38 @@
+    return this._theme;
+  }
+
+  public subscribeOnThemeChange(cb: (next: Theme) => void): Subscription {
```
nit: next -> newTheme or just "theme"

MahmoudAGawad commented on diagnostics-app/src/app/layout/header/header.component.it.spec.ts:
```diff
@@ -0,0 +1,60 @@
+// found in the LICENSE file.
+
+/**
+ * @fileoverview Integration Tests for header.component
```
nit: Integration Tests -> Integration tests

MahmoudAGawad commented on diagnostics-app/src/app/layout/header/header.component.unit.spec.ts:
```diff
@@ -0,0 +1,45 @@
+// found in the LICENSE file.
+
+/**
+ * @fileoverview Unit Tests for header.component
```
nit: Unit Tests -> Unit tests

MahmoudAGawad reviewed (approved):
Thank you for the effort! Looks much better now.

LGTM % nits.

After resolving the nits, please resolve the open conversations and go ahead and merge.

Thanks!
